app/src/main/java/com/celzero/bravedns/ui/SettingsFragment.kt

```diff
@@ -659,78 +701,73 @@ class SettingsFragment : Fragment(R.layout.activity_settings_screen) {
+                    val timeStamp = jsonObject.getLong(Constants.JSON_LATEST)
+                    if (DEBUG) Log.d(LOG_TAG_DOWNLOAD, "onResponse -  $updateValue")
+                    if (updateValue) {
+                        persistentState.tempBlocklistDownloadTime = timeStamp
+                        persistentState.workManagerStartTime = SystemClock.elapsedRealtime()
+                        get<AppDownloadManager>().downloadLocalBlocklist(timeStamp)
+                        updateDownloadInitiated()
+                        return
+                    }
+                    if (!isUserInitiated) {
                         updateDownloadFailure()
+                        return
                     }
+                    if (isVpnLockdownEnabled() == true) {
+                        showRedownloadDialogLockdown(timeStamp)
+                    } else {
+                        showRedownloadDialog(timeStamp)
+                    }
+                } catch (e: java.lang.Exception) {
+                    Log.w(LOG_TAG_DOWNLOAD, "Exception while downloading: ${e.message}", e)
+                    updateDownloadFailure()
                 }
             }
         })
     }
 
-    private fun updateDownloadFailure(){
-        b.settingsActivityOnDeviceBlockConfigureBtn.visibility = View.GONE
-        b.settingsActivityOnDeviceLastUpdatedTimeTxt.visibility = View.GONE
-        b.settingsActivityOnDeviceBlockRefreshBtn.visibility = View.GONE
-        b.settingsActivityOnDeviceBlockProgress.visibility = View.GONE
-        b.settingsActivityOnDeviceBlockSwitch.visibility = View.VISIBLE
-        b.settingsActivityOnDeviceBlockSwitch.isChecked = false
-        b.settingsActivityOnDeviceBlockDesc.text = getString(R.string.settings_local_blocklist_desc4)
-        persistentState.workManagerStartTime = 0
+    private fun onDownloadCheckFailure(isRetry: Boolean) {
+        activity?.runOnUiThread {
+            if (isRetry) {
+                Utilities.showToastUiCentered(requireContext(), getString(R.string.local_blocklist_update_check_failure), Toast.LENGTH_SHORT)
+            } else {
+                updateDownloadFailure()
+            }
+        }
     }
 
-    private fun updateDownloadSuccess(){
+    private fun updateDownloadFailure() {
+        activity?.runOnUiThread {
+            b.settingsActivityOnDeviceBlockConfigureBtn.visibility = View.GONE
+            b.settingsActivityOnDeviceLastUpdatedTimeTxt.visibility = View.GONE
+            b.settingsActivityOnDeviceBlockRefreshBtn.visibility = View.GONE
+            b.settingsActivityOnDeviceBlockProgress.visibility = View.GONE
+            b.settingsActivityOnDeviceBlockSwitch.visibility = View.VISIBLE
+            b.settingsActivityOnDeviceBlockSwitch.isChecked = false
+            b.settingsActivityOnDeviceBlockDesc.text = getString(R.string.settings_local_blocklist_desc4)
+            persistentState.workManagerStartTime = 0
```

Comment:
minor: there a `init-time` constant defined, right, for use with zero-time?

Reply:
Assigned const value `INIT_TIME_MS` for `persistentState.workManagerStartTime`